src/Orleans.Core/Streams/QueueAdapters/BatchContainerBatch.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// A batch of batch containers, that if configured (see StreamPullingAgentOptions), will be the data pulled by the
+    /// PersistenStreamPullingAgent from it's underlying cache
+    /// </summary>
+    class BatchContainerBatch : IBatchContainerBatch
+    {
+        /// <summary>
+        /// Stream identifier for the stream this batch is part of.
+        /// Derived from the first batch container in the batch.
+        /// </summary>
+        public Guid StreamGuid { get; }
+
+        /// <summary>
+        /// Stream Sequence Token for the start of this batch.
+        /// Derived from the first batch container in the batch.
+        /// </summary>
+        public StreamSequenceToken SequenceToken { get; }
+
+        /// <summary>
+        /// Stream namespace for the stream this batch is part of.
+        /// Derived from the first batch container in the batch.
+        /// </summary>
+        public string StreamNamespace { get; }
+
+        /// <summary>
+        /// Batch containers comprising this batch
+        /// </summary>
+        public IEnumerable<IBatchContainer> BatchContainers { get; }
+
+        public BatchContainerBatch(IEnumerable<IBatchContainer> batchContainers)
+        {
+            if ((batchContainers == null) || !batchContainers.Any())
+            {
+                throw new ArgumentNullException(nameof(batchContainers));
+            }
+
+            this.BatchContainers = batchContainers;
+
+            var containerDelegate = this.BatchContainers.First();
+            this.SequenceToken = containerDelegate.SequenceToken;
+            this.StreamGuid = containerDelegate.StreamGuid;
+            this.StreamNamespace = containerDelegate.StreamNamespace;
+        }
+
+        public IEnumerable<Tuple<T, StreamSequenceToken>> GetEvents<T>()
+        {
+            return this.BatchContainers.SelectMany(batchContainer => batchContainer.GetEvents<T>());
+        }
+
+        public bool ImportRequestContext()
+        {
+            return true;
```

Comment:
Need to set call import context on each batch prior to returning it's events in GetEvents.  Suggest returning events via yield loop in GetEvents to support this.

Reply:
Are you saying that within GetEvents itself I should call ImportRequestContext on each batch and then yield that batch?

Reply:
Yes, or something of the like.  The expected behavior is that when events from a batch are passed to the application, they are done so within the calling context of the batch they are in.  This may not be possible for batched observers.  I'll speak to the team about that behavior.

Reply:
Done. Added import context in StreamSubscriptionHandleImpl when iterating over the IBatchContainers which compose the IBatchContainerBatch.